felix/calc/active_bpgpeer_calculator.go

```diff
@@ -0,0 +1,299 @@
+
+	// Active BGP peers.
+	bgpPeersByName map[string]*v3.BGPPeer
+
+	// Local Workload selectors of active BGP peers.
+	workloadSelectorsByPeerName map[string]sel.Selector
+
+	// peer data of active local endpoints.
+	peersByID map[model.WorkloadEndpointKey]EpPeerData
+
+	// Callbacks.
+	OnEndpointBGPPeerDataUpdate func(id model.WorkloadEndpointKey, peerData *EpPeerData)
+}
+
+// Peer information that we track for each active local endpoint.
+type EpPeerData struct {
+	// Name of the V3 BGPPeer resource.
+	v3PeerName string
+}
+
+// Return true if the peer data is empty.
+func (p EpPeerData) isEmpty() bool {
+	return len(p.v3PeerName) == 0
+}
+
+// Return true if the peer data is associated with a BGPPeer resource.
+func (p EpPeerData) associatedWith(name string) bool {
+	return p.v3PeerName == name
+}
+
+func NewActiveBGPPeerCalculator(hostname string) *ActiveBGPPeerCalculator {
+	abp := &ActiveBGPPeerCalculator{
+		hostname:                    hostname,
+		nodeLabels:                  map[string]string{},
+		labelsByID:                  map[model.WorkloadEndpointKey]map[string]string{},
+		workloadSelectorsByPeerName: map[string]sel.Selector{},
+		bgpPeersByName:              map[string]*v3.BGPPeer{},
+		peersByID:                   map[model.WorkloadEndpointKey]EpPeerData{},
+	}
+	return abp
+}
+
+func (abp *ActiveBGPPeerCalculator) RegisterWith(localEndpointDispatcher, allUpdDispatcher *dispatcher.Dispatcher) {
+	// It needs local workload endpoints.
+	localEndpointDispatcher.Register(model.WorkloadEndpointKey{}, abp.OnUpdate)
+	// It also needs Nodes and BGPPeers.
+	allUpdDispatcher.Register(model.ResourceKey{}, abp.OnUpdate)
+}
+
+func (abp *ActiveBGPPeerCalculator) OnUpdate(update api.Update) (_ bool) {
+	logCxt := logrus.WithField("update", update)
+	switch id := update.Key.(type) {
+	case model.WorkloadEndpointKey:
+		if update.Value != nil {
+			logCxt.Debug("Updating abp with endpoint update")
+			endpoint := update.Value.(*model.WorkloadEndpoint)
+			abp.onEndpointUpdate(id, endpoint.Labels)
+		} else {
+			logCxt.Debug("Deleting endpoint from abp")
+			abp.onEndpointDelete(id)
+		}
+	case model.ResourceKey:
+		switch id.Kind {
+		case v3.KindBGPPeer:
+			if update.Value != nil {
+				logCxt.Info("Updating abp with BGPPeer")
+				bgpPeer := update.Value.(*v3.BGPPeer)
+
+				name := bgpPeer.Name
+				// Save latest bgpPeer.
+				abp.bgpPeersByName[name] = bgpPeer
+
+				if !abp.ifBgpPeerSelectHost(bgpPeer) {
+					// Trying to delete BGPPeer if it does not select the host.
+					abp.onBGPPeerDelete(name)
+				} else {
+					abp.onBGPPeerUpdate(bgpPeer)
+				}
+			} else {
+				logCxt.Debug("Deleting BGPPeer from abp")
+				abp.onBGPPeerDelete(id.Name)
+				delete(abp.bgpPeersByName, id.Name)
+			}
+		case libv3.KindNode:
+			if update.Value != nil {
+				nodeName := update.Key.(model.ResourceKey).Name
+
+				logCxt := logrus.WithField("node", nodeName)
+
+				if nodeName == abp.hostname {
```

Comment:
```suggestion
				nodeName := update.Key.(model.ResourceKey).Name

				if nodeName != abp.hostname {
				    return
				}
				logCxt := logrus.WithField("node", nodeName)

```
Go idiom is to return early rather than nest more deeply.